### PR TITLE
Add cosmos genesis tinker script with example

### LINKED ISSUE #28: Revise genesis modification script that is able to read values from an exported genesis file and determine new values that don't break invariants
- [ ] udit to set up pairing time with el/bae

### PR DESCRIPTION
**Pull request by RangerMauve**
Adds the initial version of the cosmos genesis tinker script.

This can swap out delegator keys, change balances, increase validator power, and create tokens.

Closes #28

### CODE REVIEW

**Comment by uditvira** [comment]
I have generated a genesis from our mainnet validator. I'm going to try using this script to modify that genesis file.

File: cosmos-genesis-tinker/cosmos_genesis_tinker.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+import json
+from hashlib import sha256
+from zipfile import ZipFile
+from io import BytesIO
+import gzip
+import tarfile
+import requests
+
+BINANCE_VALIDATOR_ADDRESS = "cosmosvaloper156gqf9837u7d4c4678yt3rl4ls9c5vuursrrzf"
```
Comment (uditvira) [inline cosmos-genesis-tinker/cosmos_genesis_tinker.py:16]: I don't think is required any more!

File: cosmos-genesis-tinker/cosmos_genesis_tinker.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        for info in starting_infos:
+            if info["delegator_address"] == old_address:
+                info["delegator_address"] = new_address
+                break
+
+        return self
+
+    def create_coin(self, denom, amount="0"):
+        """
+        Creats a new coin based on a given name if it doesn't exist
```
Comment (uditvira) [inline cosmos-genesis-tinker/cosmos_genesis_tinker.py:211]:
```suggestion
        Creates a new coin based on a given name if it doesn't exist
```

File: cosmos-genesis-tinker/cosmos_genesis_tinker.py (hunk 1 above)
Comment (uditvira) [reply cosmos-genesis-tinker/cosmos_genesis_tinker.py:16]:
```suggestion
```

File: cosmos-genesis-tinker/cosmos_genesis_tinker.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+import json
+from hashlib import sha256
+from zipfile import ZipFile
+from io import BytesIO
+import gzip
+import tarfile
+import requests
+
+BINANCE_VALIDATOR_ADDRESS = "cosmosvaloper156gqf9837u7d4c4678yt3rl4ls9c5vuursrrzf"
+BINANCE_TOKEN_BONDING_POOL_ADDRESS = "cosmos1fl48vsnmsdzcv85q5d2q4z5ajdha8yu34mf0eh"
```
Comment (uditvira) [inline cosmos-genesis-tinker/cosmos_genesis_tinker.py:17]: This is actually not a Binance specific address, but is the Module Account for the staking module to keep track of all bonded tokens in the Cosmos Hub (chain id "cosmoshub-4")
```suggestion
# Default value is the cosmoshub-4 bonded token pool account
TOKEN_BONDING_POOL_ADDRESS = "cosmos1fl48vsnmsdzcv85q5d2q4z5ajdha8yu34mf0eh"
```

File: cosmos-genesis-tinker/cosmos_genesis_tinker.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+        return sha256(content_bytes).hexdigest()
+
+    def save_file(self, path):
+        """
+        Save a modified genesis file back to JSON
+        """
+        with open(path, "w", encoding="utf8") as file:
+            json.dump(self.genesis, file, indent=False)
+        return self
+
```
Comment (uditvira) [inline cosmos-genesis-tinker/cosmos_genesis_tinker.py:95]: realizing there is a missing API `def change_chain_id(self, new_chain_id)`

File: cosmos-genesis-tinker/cosmos_genesis_tinker.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+                new_amount = old_amount + increase
+                coin["amount"] = str(new_amount)
+                found_coin = True
+                break
+
+        if not found_coin:
+            self.create_coin(denom, increase)
+
+        return self
+
```
Comment (uditvira) [inline cosmos-genesis-tinker/cosmos_genesis_tinker.py:275]: There are a few linked changes, so I think they should be wrapped in an API, something like:

```suggestion
def increase_delegator_stake_to_validator(self, delegator, validator, stake, token_bonding_pool_address=TOKEN_BONDING_POOL_ADDRESS):
  self.increase_balance(token_bonding_pool_address, stake)
  self.increase_delegator_stake(delegator, stake)
  self.increase_validator_stake(validator, stake)
  self.increase_validator_power(validator, stake)
```

**Review by uditvira — CHANGES_REQUESTED** [review]

**Comment by uditvira** [comment]
Also I think a really nice feature to have would be to log all the changes that have been made to stdout. Would help to be able to copy-paste them into a github issue for example

**Comment by RangerMauve** [comment]
> Also I think a really nice feature to have would be to log all the changes that have been made to stdout. Would help to be able to copy-paste them into a github issue for example

What sort of output were you expecting in the STDOUT? Like, the function calls that are being invoked?

**Comment by uditvira** [comment]
> What sort of output were you expecting in the STDOUT? Like, the function calls that are being invoked?

imagining a console dump like

````
Summary of changes
1. Changed chain-id to theta-testnet
2. Increased balance of cosmosxxxxxd23 by 1234 ATOM
3. Replaced validator cosmosvalopasfsdf with comsosvalopasdads
4. ...
```

**Comment by RangerMauve** [comment]
After adding the new method, I'm getting the following error:

```
Traceback (most recent call last):
  File "/home/mauve/programming/cosmos-organizing/cosmos-genesis-tinker/./genesis-example.py", line 82, in <module>
    genesis.increase_delegator_stake_to_validator(
  File "/home/mauve/programming/cosmos-organizing/cosmos-genesis-tinker/cosmos_genesis_tinker.py", line 415, in increase_delegator_stake_to_validator
    self.increase_validator_stake(validator, stake)
  File "/home/mauve/programming/cosmos-organizing/cosmos-genesis-tinker/cosmos_genesis_tinker.py", line 378, in increase_validator_stake
    raise Exception("Could not find operator_address")
Exception: Could not find operator_address
```

The change is that before genesis-example was using the binance validator address for stake (I think incorrectly?) and now it's (correctly?) using `NODE2_NEW_ADDRESS`, but it seems the address isn't being found. Is there something going wrong in the swap?

**Comment by uditvira** [comment]
> After adding the new method, I'm getting the following error:
> 
> ```
> Traceback (most recent call last):
>   File "/home/mauve/programming/cosmos-organizing/cosmos-genesis-tinker/./genesis-example.py", line 82, in <module>
>     genesis.increase_delegator_stake_to_validator(
>   File "/home/mauve/programming/cosmos-organizing/cosmos-genesis-tinker/cosmos_genesis_tinker.py", line 415, in increase_delegator_stake_to_validator
>     self.increase_validator_stake(validator, stake)
>   File "/home/mauve/programming/cosmos-organizing/cosmos-genesis-tinker/cosmos_genesis_tinker.py", line 378, in increase_validator_stake
>     raise Exception("Could not find operator_address")
> Exception: Could not find operator_address
> ```
> 
> The change is that before genesis-example was using the binance validator address for stake (I think incorrectly?) and now it's (correctly?) using `NODE2_NEW_ADDRESS`, but it seems the address isn't being found. Is there something going wrong in the swap?

This issue was because we were passing an operator address instead of a node address by mistake. I fixed here: https://github.com/hyphacoop/cosmos-organizing/pull/55/commits/54d276232aac7adef18302a20e7603a642ae2486

**Comment by uditvira** [comment]
I had to make significant changes here in the increase_validator_power function https://github.com/hyphacoop/cosmos-organizing/pull/55/commits/dca7adb97dde939fec508edb60ef541391741b1d This is because we need to deal with a case where the validator address isn't in the validators list if the validator has less voting power than the top 150. This is the case for the modifications we are doing for the theta testnet

**Comment by uditvira** [comment]
We also need a way to modify the following governance parameters:
* quorum
* min deposit amount 
* threshold
* voting period

You'll find these all under `self.genesis['app_state']['gov']`. This is what the json looks like:
```
"deposit_params": {
"max_deposit_period": "1209600s",
"min_deposit": [
{
"amount": "64000000",
"denom": "uatom
}
```

```
"tally_params": {
"quorum": "0.400000000000000000",
"threshold": "0.500000000000000000",
"veto_threshold": "0.334000000000000000"
}
```

```
"voting_params": {
"voting_period": "1209600s"
}
```
See here: https://github.com/cosmos/vega-test/tree/master/local-testnet#modify-some-gov-parameters-to-improve-testing-efficiency

**Comment by RangerMauve** [comment]
Closing this in favor of the new repo: https://github.com/hyphacoop/cosmos-genesis-tinkerer/pull/1

**Pull request closed without merging**